mava/systems/sable/anakin/ff_sable.py

```diff
@@ -0,0 +1,664 @@
+
+from mava.evaluator import ActorState, EvalActFn, get_eval_fn, get_num_eval_envs
+from mava.networks import SableNetwork
+from mava.networks.utils.sable import get_init_hidden_state
+from mava.systems.sable.types import (
+    ExecutionApply,
+    TrainingApply,
+    Transition,
+)
+from mava.systems.sable.types import FFLearnerState as LearnerState
+from mava.types import Action, ExperimentOutput, LearnerFn, MarlEnv
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax_utils import merge_leading_dims, unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.utils.network_utils import get_action_head
+from mava.utils.total_timestep_checker import check_total_timesteps
+from mava.utils.training import make_learning_rate
+from mava.wrappers.episode_metrics import get_final_step_metrics
+
+
+def get_learner_fn(
+    env: Environment,
+    apply_fns: Tuple[ExecutionApply, TrainingApply],
+    update_fn: optax.TransformUpdateFn,
+    config: DictConfig,
+) -> LearnerFn[LearnerState]:
+    """Get the learner function."""
+
+    # Get apply functions for executing and training the network.
+    sable_action_select_fn, sable_apply_fn = apply_fns
+
+    def _update_step(learner_state: LearnerState, _: Any) -> Tuple[LearnerState, Tuple]:
+        """A single update of the network.
+
+        This function steps the environment and records the trajectory batch for
+        training. It then calculates advantages and targets based on the recorded
+        trajectory and updates the actor and critic networks based on the calculated
+        losses.
+
+        Args:
+        ----
+            learner_state (NamedTuple):
+                - params (FrozenDict): The current model parameters.
+                - opt_states (OptState): The current optimizer states.
+                - key (PRNGKey): The random number generator state.
+                - env_state (State): The environment state.
+                - last_timestep (TimeStep): The last timestep in the current trajectory.
+            _ (Any): The current metrics info.
+
+        """
+
+        def _env_step(learner_state: LearnerState, _: int) -> Tuple[LearnerState, Transition]:
+            """Step the environment."""
+            params, opt_states, key, env_state, last_timestep = learner_state
+
+            # SELECT ACTION
+            key, policy_key = jax.random.split(key)
+
+            # Apply the actor network to get the action, log_prob, value and updated hstates.
+            last_obs = last_timestep.observation
+            action, log_prob, value, _ = sable_action_select_fn(  # type: ignore
+                params,
+                obs_carry=last_obs,
+                key=policy_key,
+            )
+            action = jnp.squeeze(action, axis=-1)
+            log_prob = jnp.squeeze(log_prob, axis=-1)
+            value = jnp.squeeze(value, axis=-1)
```

Comment:
Can we handle all the squeezing inside the networks if possible? This makes supporting continuous actions easier later on as well.